src/raftstore/coprocessor/split_observer.rs

```diff
@@ -48,14 +53,58 @@ impl SplitObserver {
             key.truncate(table::PREFIX_LEN + table::ID_LEN);
         }
 
-        let region_start_key = ctx.region().get_start_key();
-
         let key = encode_bytes(&key);
-        if *key <= *region_start_key {
-            return Err("no need to split".to_owned());
+        match util::check_key_in_region_exclusive(&key, region) {
+            Ok(()) => Ok(key),
+            Err(_) => Err(format!(
+                "key \"{}\" should be in (\"{}\", \"{}\")",
+                escape(&key),
+                escape(region.get_start_key()),
+                escape(region.get_end_key()),
+            )),
         }
+    }
 
-        split.set_split_key(key);
+    fn on_split(&self, ctx: &mut ObserverContext, splits: &mut Vec<SplitRequest>) -> Result<()> {
+        let (mut i, mut j) = (0, 0);
+        let mut last_valid_key: Option<Vec<u8>> = None;
+        let region_id = ctx.region().get_id();
+        while i < splits.len() {
+            let k = i;
+            i += 1;
+            {
+                let split = &mut splits[k];
+                let key = split.take_split_key();
+                match self.adjust_key(ctx.region(), key) {
+                    Ok(key) => {
+                        if last_valid_key.as_ref().map_or(false, |k| *k >= key) {
```

Comment:
I think `k >= &key` is better

Reply:
No, `*` is preferred.

Reply:
I don't quite understand, wouldn't `*k` cause copy?

Reply:
I don't quite understand, wouldn't *k cause copy?

Reply:
No, `*` or `&` is the same in this case.